Better handling of multiple rf system

swhite2401 opened the pull request:
**energy_loss.set_cavity_phase()**

In case multiple RF systems are found a least_square minimization is used to find the `TimeLag` from `U0` and the sum of the sine waves. A single TimeLag is applied to all RF systems.
This method is called only is multiple RF frequencies are found in `cavpts`

**orbit.find_orbit6()**

Instead of using the first cavity the determine the harmonic number and rf frequency, the following method is used:
1-look for a unique rf frequency using `cavity_access()`
2-If 1 fails, i.e. multiple rf frequencies are found, the cavity with the smallest frequency is used as reference
The initial guess was sometimes failing for non-zero `TimeLag` as it was not taken into account. 
Presently the only function setting the TimeLag is set_cavity_phase() which sets the initial point to zero. If a non-zero `TimeLag` is found  `numpy.zeros(6)` is used as initial guess, 
To be noted, in case the `TimeLag` is not set and multiple RF systems are present, the initial guess is forced to `numpy.zeros(6)` and it is possible that `find_orbit6()` does not converge.

lfarv commented:
#### set_cavity_phase
No ! You are putting all cavities in phase ! The 1st RF goals I see (harmonic bunch-lengthening, fs modulation, bunch splitting/grouping…) all require an adequate phasing between the cavities. Putting them in phase ruins the goal ! What you have to do is:
- include the initial `TimeLag` in the equation (`sin(2*pi*freq*(x-timelag)/clight)`),
- **decrement all** `TimeLag`s with the solution.

This way, you shift the time origin to bring the synchronous phase to zero, which is the goal of `set_cavity_phase`, but keep the desired voltage shape. If for some particular purpose you need cavities in phase, they will stay in phase after running `set_cavity_phase`, so no problem.
There is still a "detail" to solve: the least square fit has 50% chance to find an unstable fixed point. You have to choose a solution with the right voltage slope.

#### find_orbit6
It is much simpler: now that you can solve the equation `sum( Vi(t_sync)) = U0`, you just set the initial guess to `t_sync` and you are done (provide you take a stable solution). No test, no choice. Just for efficiency, in the case of a single system, you can use the analytical solution (the present one) rather than the least squares.

However, you should keep the accessibility of the `guess` keyword, it's the only way to select the desired bucket, since in a complex system, you may have several buckets with different properties.

swhite2401 commented:
Thanks for the feedback, in fact your are taking a step further, I just had in mind to facilitate `find_orbit6` convergence.
I am a bit busy these days but will implement your proposals as soon as I can!

swhite2401 commented:
Corrected version:

**set_cavity_phase**
-least_square used to find the zero crossing of the derivative
-> use the side with positive slope to set the boundaries for the least_square minimization to find the synchronous phase
-The initial TimeLag is handled as advised for complex RF systems

**find_orbit6**
-Simplified version according to previous comments

I had to relax a bit the tolerance on test_ohmi_envelop for the comparison of emittances computed
with matlab, probably due to the different solution found by find_orbit6

I am not fully satisfied with this double call to least_square....any advices?

I hope I got all the signs rights...from the test I have done it seems to work fine but a double check would be good.

lfarv commented on pyat/at/physics/energy_loss.py:
```diff
@@ -81,6 +84,59 @@ def tracking(ring):
+            raise AtError('Not enough RF voltage: unstable ring')
+        ctmax = 1/numpy.amin(freq)*clight/2
+        tt0 = tl0[numpy.argmin(freq)]
+        eq = lambda x: numpy.sum(rfv*numpy.sin(2*pi*freq*(x+tl0)/clight))-u0
+        deq = lambda x: numpy.sum(rfv*numpy.cos(2*pi*freq*(x+tl0)/clight))
```
`CavityPass` uses a minus sign: `sin(2*pi*freq*(ct - timelag)/clight))`. Should not you use the same convention?

swhite2401 replied:
The sign is consistent with this: `timelag = clight/(2*pi*freq)*numpy.arcsin(u0/rfv)`

lfarv commented on pyat/at/physics/energy_loss.py:
```diff
@@ -104,16 +160,13 @@ def set_cavity_phase(ring, method=ELossMethod.INTEGRAL,
         cavpts = refpts
-    rfv = ring.get_rf_voltage(cavpts=cavpts)
-    freq = ring.get_rf_frequency(cavpts=cavpts)
+    elif cavpts is None:
+        cavpts = get_refpts(ring, RFCavity)
```
Redundant test: caveats if properly handled in both `get_timelag_fromU0` and `set_value_refpts`

swhite2401 replied:
set_value_refpts does not handle refpts=None, I am not I understand what you mean

lfarv commented on pyat/at/physics/energy_loss.py:
```diff
@@ -81,6 +84,59 @@ def tracking(ring):
+    else:
+        if u0 > rfv:
+            raise AtError('Not enough RF voltage: unstable ring')
+        timelag = clight/(2*pi*freq)*numpy.arcsin(u0/rfv) * \
```
`cavpts` may be a boolean mask or a test function. So `len(cavpts)` is not always the number of refpoints. You can use `refpts_len(ring,refpts)` or `ring.refcount(refpts)` to get the number. Or better, keep it as a scalar, it should be broadcast to the right size when necessary.

swhite2401 replied:
Changed to get_cells
I don't like that the function can return different "types" depending on the case so I prefer to leave it as is and return always a scalar

lfarv commented on pyat/at/physics/energy_loss.py:
```diff
@@ -81,6 +84,59 @@ def tracking(ring):
+            raise AtError('Not enough RF voltage: unstable ring')
+        timelag = clight/(2*pi*freq)*numpy.arcsin(u0/rfv) * \
+                  numpy.ones(len(cavpts))
+    return timelag, numpy.squeeze(numpy.unique(timelag-tl0))
```
You could remove the expensive call to "`unique`" by defining `ts = timelag-tl0` in the "simple case, and returning it

swhite2401 replied:
Done

lfarv commented:
Now I need some time to look at the solution of the equation itself, more on Monday evening…

swhite2401 commented:
Ok to merge?

lfarv reviewed (approved):
Ok for now, we'll see how it works.
